### PR TITLE
feat: Introduce "minimum supported Client version" configuration parameter for Server Proxy

### PR DESCRIPTION
**Pull request by andresrosenthal**
Refs: XRDDEV-1999

### CODE REVIEW

**Review by enelir — APPROVED** [review]

File: src/proxy/src/main/java/ee/ria/xroad/proxy/serverproxy/ClientProxyVersionCheck.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import org.semver4j.Semver;
+
+import javax.servlet.http.HttpServletRequest;
+
+import java.util.Optional;
+
+import static ee.ria.xroad.common.ErrorCodes.X_CLIENT_PROXY_VERSION_NOT_SUPPORTED;
+
+@Slf4j
+public class ClientProxyVersionCheck {
```
Comment (andresrosenthal) [inline src/proxy/src/main/java/ee/ria/xroad/proxy/serverproxy/ClientProxyVersionCheck.java:45]:
```suggestion
public class ClientProxyVersionVerifier {
```

File: src/proxy/src/main/java/ee/ria/xroad/proxy/serverproxy/ClientProxyVersionCheck.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    private static final String UNKNOWN_VERSION = "unknown";
+    private static final Semver MIN_SUPPORTED_CLIENT_VERSION;
+
+    static {
+        MIN_SUPPORTED_CLIENT_VERSION = Optional.ofNullable(SystemProperties.getServerProxyMinSupportedClientVersion())
+                .map(Semver::new)
+                .orElse(null);
+    }
+
+    void check(HttpServletRequest request) {
```
Comment (andresrosenthal) [inline src/proxy/src/main/java/ee/ria/xroad/proxy/serverproxy/ClientProxyVersionCheck.java:55]: As this class does seem like an utility class I think this method should should be made static.
Note that then there's no need to inject an instance of this class to `ServerProxyHandler`

File: src/proxy/src/main/java/ee/ria/xroad/proxy/serverproxy/ClientProxyVersionCheck.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        String clientVersion = getVersion(request.getHeader(MimeUtils.HEADER_PROXY_VERSION));
+        String thisVersion = getVersion(ProxyMain.readProxyVersion());
```
Comment (andresrosenthal) [inline src/proxy/src/main/java/ee/ria/xroad/proxy/serverproxy/ClientProxyVersionCheck.java:57]: This line could be moved to the line just before the ```if (!clientVersion.equals(thisVersion)) {``` check, where `thisVersion` variable is used for the first time.

File: src/proxy/src/test/java/ee/ria/xroad/proxy/serverproxy/ClientProxyVersionCheckTest.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    @Test
+    public void shouldRaiseError() {
+        System.setProperty(MIN_SUPPORTED_CLIENT_VERSION, VERSION_7_1_3);
+        final HttpServletRequest request = mock(HttpServletRequest.class);
+        when(request.getHeader(MimeUtils.HEADER_PROXY_VERSION)).thenReturn(CLIENT_VERSION_6_26_3);
+        clientProxyVersionCheck = new ClientProxyVersionCheck();
+
+        assertThrows(CodedException.class, () -> clientProxyVersionCheck.check(request));
```
Comment (andresrosenthal) [inline src/proxy/src/test/java/ee/ria/xroad/proxy/serverproxy/ClientProxyVersionCheckTest.java:72]: Could you please also assert exception's message?

**Pull request merged**